kge/job/train.py

```diff
@@ -583,55 +596,65 @@ def collate(batch):
                 if example_index < num_sp:
                     num_ones += self.train_sp_offsets[example_index + 1]
                     num_ones -= self.train_sp_offsets[example_index]
-                else:
+                elif example_index < end_po:
                     example_index -= num_sp
                     num_ones += self.train_po_offsets[example_index + 1]
                     num_ones -= self.train_po_offsets[example_index]
+                else:
+                    example_index -= end_po
+                    num_ones += self.train_so_offsets[example_index + 1]
+                    num_ones -= self.train_so_offsets[example_index]
 
             # now create the results
-            sp_po_batch = torch.zeros([len(batch), 2], dtype=torch.long)
-            is_sp = torch.zeros([len(batch)], dtype=torch.long)
+            sp_po_so_batch = torch.zeros([len(batch), 2], dtype=torch.long)
+            key_type = torch.zeros([len(batch)], dtype=torch.long)
             label_coords = torch.zeros([num_ones, 2], dtype=torch.int)
             current_index = 0
             triples = torch.zeros([num_ones, 3], dtype=torch.long)
             for batch_index, example_index in enumerate(batch):
-                is_sp[batch_index] = 1 if example_index < num_sp else 0
-                if is_sp[batch_index]:
+                key_type[batch_index] = 1 if example_index < num_sp else (0 if example_index < end_po else 2)
+                if key_type[batch_index] == 1:
                     keys = self.train_sp_keys
                     offsets = self.train_sp_offsets
                     values = self.train_sp_values
-                    sp_po_col_1, sp_po_col_2, o_s_col = S, P, O
-                else:
+                    sp_po_so_col_1, sp_po_so_col_2, o_s_p_col = S, P, O
```

Comment:
try to use more intuitive variable names: query_col1, query_col2, target_col for example